src/NuGet.Core/NuGet.Protocol/Resources/ReadmeUriTemplateResource.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+#if NETCOREAPP
+using System;
+#endif
+
+namespace NuGet.Protocol
+{
+    internal class ReadmeUriTemplateResource : INuGetResource
+    {
+        private readonly string _uriTemplate;
+
+        public ReadmeUriTemplateResource(string uriTemplate)
+        {
+            _uriTemplate = uriTemplate;
+        }
+
+        /// <summary>
+        /// Get  the URL for downloading the readme file.
+        /// </summary>
+        /// <param name="id">The package id</param>
+        /// <param name="version">The package version</param>
+        /// <returns>URL to download README, built using the URI template.</returns>
+        public string GetReadmeUrl(string id, NuGetVersion version)
+        {
+            if (_uriTemplate == null)
+            {
+                return string.Empty;
+            }
+
+            var uriString = _uriTemplate
+#if NETCOREAPP
+               .Replace("{lower_id}", id.ToLowerInvariant(), StringComparison.OrdinalIgnoreCase)
```

Comment:
Could you help me understand the lower-casing of ID here? I don't think protocol expects these to be lowercase based on https://learn.microsoft.com/en-us/nuget/api/package-details-template-resource#url-placeholders. I might be misunderstanding that doc.

Reply:
Here's an example of one that excepts it to be lower. https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource. @joelverhagen it doesn't seem like the case is important for downloading the README from nuget.org.

Reply:
Look forward to hearing Joel's clarification. To restate my concern, I wouldn't want us putting something in Client that's specific to nuget.org. Since ID is globally unique, passing lowercase values in the query string sounds like a server implementation detail to me.

Reply:
> Here's an example of one that excepts it to be lower. https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource. @joelverhagen it doesn't seem like the case is important for downloading the README from nuget.org.

Hmm never mind, it does look like the NuGet.org URL is case sensitive. 
I had originally suggested we provide options for lower_id and upper_id. We don't do that for any of the other APIs though. [Package metadata](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource) and [package content](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) both use the lower_id.

I think passing a lower ID is likely the safest choice. I imagine for the most part servers that implement these endpoints make them case insensitive. If they are going to implement something that is case sensitive then it really only makes sense to do upper or lower, since package IDs don't follow a specific casing, https://www.nuget.org/packages/NuGet.Frameworks, for example.

Reply:
Given this is new protocol we can define it however we want. The NuGet.org implementation needs lowercase.

In other places as @jgonz120 mentioned, V3 protocol documents lower ID and lower normalized version to be used. This should be the same.

Generally speaking URLs are not case insensitive but it is also common for web services to choose to be case insensitive. This is easy to do when you have a web app under your control handling the request. It's really a server to service decision. We can say "lower case IDs before putting them in the URL" for NuGet protocol. This gives predictability for case sensitive underlying tech (such as blob storage) but of course also works for case insensitive implementations.

Reply:
I'm trying to think if I've ever seen a web api that's case sensitive that requires everything to be upper case. I would feel like everything is yelling at me.